### PR TITLE
Documentation for caching layers

### LINKED ISSUE #18476: Documentation for caching layers
This ticket will add documentation describing the basic purpose of the `SpectrumInfo` and `DetectorInfo` caching layers, and contain examples of the changes required for the roll out.

### PR DESCRIPTION
**Pull request by ianbush**
Documentation for the caching layers `DetectorInfo` and `SpectrumInfo` to help with the rollout.

**To test:**

Read through document.

Fixes #18476.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/concepts/CachingLayers.rst [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+Caching Layers - SpectrumInfo and DetectorInfo
+==============================================
+
+.. contents::
+  :local:
+
+Introduction
+------------
+
+There are two caching layers, ``SpectrumInfo`` and ``DetectorInfo``, that are going to be introduced to Mantid as part of the work towards Instrument 2.0. Eventually these classes will cache commonly accessed information about spectra or detectors, namely masking, monitor flags, L1, L2, 2-theta and position, leading to improved performance and cleaner code.
```
Comment (OwenArnold) [inline docs/source/concepts/CachingLayers.rst]: Good intro.

File: docs/source/concepts/CachingLayers.rst [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+____________
+
+``SpectrumInfo`` can be obtained from a call to ``MatrixWorkspace::spectrumInfo()``. The wrapper class holds a reference to a ``DetectorInfo`` object and calls through to this for access to information on masking, monitor flags etc.
+
+DetectorInfo
+____________
+
+``DetectorInfo`` can be obtained from a call to ``ExperimentInfo::detectorInfo()`` (usually this would be called on a ``MatrixWorkspace``). The wrapper class holds a reference to the parametrised instrument for retrieving the relevant information.
+
+There is also a real ``DetectorInfo`` class that has masking implemented, which the ``DetectorInfo`` holds a reference to. This is not important for the discussion of the rollout, but all masking operations are now going via the real ``DetectorInfo`` layer.
```
Comment (OwenArnold) [inline docs/source/concepts/CachingLayers.rst]: Can this be reworded? This isn't very clear.

**Comment by SimonHeybrock** [comment]
@ianbush Have you considered avoiding the term "caching layer", since both `SpectrumInfo` and `DetectorInfo` are not really intended to be caching layers (only some parts of them are, temporarily).

File: docs/source/concepts/CachingLayers.rst [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+==============================================
+
+.. contents::
+  :local:
+
+Introduction
+------------
+
+There are two caching layers, ``SpectrumInfo`` and ``DetectorInfo``, that are going to be introduced to Mantid as part of the work towards Instrument 2.0. Eventually these classes will cache commonly accessed information about spectra or detectors, namely masking, monitor flags, L1, L2, 2-theta and position, leading to improved performance and cleaner code.
+
```
Comment (SimonHeybrock) [inline docs/source/concepts/CachingLayers.rst]: Maybe add a short note, similar to this:

A spectrum corresponds to (a group of) one or more detectors. Most algorithm work with spectra and thus `SpectrumInfo` would be used. Some algorithms work on a lower level (with individual detectors) and thus `DetectorInfo` would be used.

File: docs/source/concepts/CachingLayers.rst (hunk 2 above)
Comment (SimonHeybrock) [inline docs/source/concepts/CachingLayers.rst]: in the new `Beamline` module.

File: docs/source/concepts/CachingLayers.rst [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+**After - check it has some detectors**
+
+.. code-block:: c++
+
+  #include "MantidAPI/SpectrumInfo.h"
+
+  ...
+
+  const aut &spectrumInfo = ws->spectrumInfo();
```
Comment (SimonHeybrock) [inline docs/source/concepts/CachingLayers.rst]: `auto`

File: docs/source/concepts/CachingLayers.rst [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+**After - check for a unique detector**
+
+.. code-block:: c++
+
+  #include "MantidAPI/SpectrumInfo.h"
+
+  ...
+
+  const aut &spectrumInfo = ws->spectrumInfo();
```
Comment (SimonHeybrock) [inline docs/source/concepts/CachingLayers.rst]: `auto`

File: docs/source/concepts/CachingLayers.rst [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+**After - looping over index**
+
+.. code-block:: c++
+
+  #include "MantidAPI/Detector.h"
+
+  ...
+
+  const auto &instrument = ws->detectorInfo();
```
Comment (SimonHeybrock) [inline docs/source/concepts/CachingLayers.rst]: `detectorInfo`

File: docs/source/concepts/CachingLayers.rst [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+**After - looping over detector IDs**
+
+.. code-block:: c++
+
+  #include "MantidAPI/Detector.h"
+
+  ...
+
+  const auto &instrument = ws->detectorInfo();
```
Comment (SimonHeybrock) [inline docs/source/concepts/CachingLayers.rst]: `detectorInfo`

**Comment by SimonHeybrock** [comment]
Latest changes look good. @OwenArnold do you have more comments or can we ship it?

**Comment by ianbush** [comment]
I just need to add something about performance tests and the link to the ticket for `DetectorInfo` rollout.

**Comment by OwenArnold** [comment]
This looks good.

**Comment by ianbush** [comment]
@SimonHeybrock @OwenArnold I have made all the changes I need to this now, so it should be ready to ship.

**Comment by SimonHeybrock** [comment]
:shipit:

File: docs/source/concepts/InstrumentAccessLayers.rst [hunk 8]
```diff
@@ -117,6 +120,11 @@ ___________
 * Get the ``SpectrumInfo`` object as a const reference and use auto - ``const auto &spectrumInfo = ws->spectrumInfo();``.
 * Do not forget to add the import - ``#include "MantidAPI/SpectrumInfo.h"``.
 
+Performance Tests
+_________________
+
+Please add performance test for any algorithms that are widely used, or might be expected to have a performance increase. See `this performance test <https://github.com/mantidproject/mantid/pull/18189/files#diff-5695221d30495359738f90b83ceb0ba3>`_ added for the ``SpectrumInfo`` rollout for an example of adding such a test.
```
Comment (OwenArnold) [inline docs/source/concepts/InstrumentAccessLayers.rst]: Could we go as far as to say that each PR should include runtime metrics for the algorithms changed? That way we could harvest them for the release notes? And also that performance tests should be added to algorithms that don't already have them?

Comment (SimonHeybrock) [inline docs/source/concepts/InstrumentAccessLayers.rst]: I like Ian's suggestion of writing performance tests only for frequently used algorithms. That would avoid spending time unless really useful.

Comment (ianbush) [inline docs/source/concepts/InstrumentAccessLayers.rst]: I have pushed some changes that hopefully emphasise the importance of this a bit more. I have left in the bit about widely used algorithms. We should just make sure when reviewing the PRs they have been added when sensible.

**Comment by OwenArnold** [comment]
I think this is good to go now. I'm just going to wait to check that the documentation jobs pass with the changes then I'll merge it.

**Pull request merged**